modules/utxo-lib/src/bitgo/Musig2.ts

```diff
@@ -0,0 +1,154 @@
+import { PsbtProprietaryKeyValueData, UtxoPsbt } from './UtxoPsbt';
+import { toPlainPublicKey, toXOnlyPublicKey } from './outputScripts';
+import { BIP32Interface } from 'bip32';
+import { musig } from '../noble_ecc';
+
+export const PSBT_IN_MUSIG_PARTICIPANT_PUB_KEYS = 'PSBT_IN_MUSIG_PARTICIPANT_PUB_KEYS';
+export const PSBT_IN_MUSIG_PARTICIPANT_PUB_KEYS_SUBTYPE = 0x00;
+
+export const PSBT_IN_TAP_MUSIG2_PUB_NONCE = 'PSBT_IN_TAP_MUSIG2_PUB_NONCE';
+export const PSBT_IN_TAP_MUSIG2_PUB_NONCE_SUBTYPE = 0x00;
+
+export function createPsbtMusig2ParticipantsKeyValData(
+  tapOutputKey: Buffer,
+  tapInternalKey: Buffer,
+  participantKeys: Buffer[]
+): PsbtProprietaryKeyValueData {
+  if (tapOutputKey.length !== 32 || tapInternalKey.length !== 32) {
+    throw new Error('Invalid tapOutputKey and tapInternalKey lengths');
+  }
+  const keydata = Buffer.concat([tapOutputKey, tapInternalKey]);
+  const value = Buffer.concat(participantKeys.map((pubkey) => toXOnlyPublicKey(pubkey)));
+  const key = {
+    identifier: PSBT_IN_MUSIG_PARTICIPANT_PUB_KEYS,
+    subtype: PSBT_IN_MUSIG_PARTICIPANT_PUB_KEYS_SUBTYPE,
+    keydata,
+  };
+  return { key, value };
+}
+
+export function createPsbtMusig2PubNonceKeyValData(
+  participantKey: Buffer,
+  tapOutputKey: Buffer,
+  pubNonces: Buffer
+): PsbtProprietaryKeyValueData {
+  if (tapOutputKey.length !== 32) {
+    throw new Error(`Invalid tapOutputKey length ${tapOutputKey.length}`);
+  }
+  if (pubNonces.length !== 66) {
+    throw new Error(`Invalid pubNonces length ${pubNonces.length}`);
+  }
+  const keydata = Buffer.alloc(96);
+  toXOnlyPublicKey(participantKey).copy(keydata);
+  tapOutputKey.copy(keydata, 32);
+  const key = {
+    identifier: PSBT_IN_TAP_MUSIG2_PUB_NONCE,
+    subtype: PSBT_IN_TAP_MUSIG2_PUB_NONCE_SUBTYPE,
+    keydata,
+  };
+  return { key, value: pubNonces };
+}
+
+export function generateMusig2Nonce(psbt: UtxoPsbt, rootWalletKey: BIP32Interface, sessionId?: Buffer): void {
+  if (sessionId && sessionId.length > Math.pow(2, 32) - 1) {
```

Comment:
The length of session ID should be 32 bytes. If provided it must either be a counter unique to this secret key (converted to an array of 32 bytes), or 32 uniformly random bytes.

Reply:
Done.